odoo/orm/fields.py

```diff
@@ -1241,24 +1225,32 @@ def update_db_related(self, model: BaseModel) -> None:
     # SQL generation methods
     #
 
-    def to_sql(self, model: BaseModel, alias: str, query: Query | None) -> SQL:
+    def to_sql(self, table: TableSQL) -> SQL:
         """ Return an :class:`SQL` object that represents the value of the given
         field from the given table alias.
 
         The query object is necessary for fields that need to add tables to the query.
         """
+        model = table._model
+        model._check_field_access(self, 'read')
         if self.compute_sql:
             if self.compute_sudo:
                 model = model.sudo()
-            sql_field = determine(self.compute_sql, model, alias, query)
+                table = table._with_model(model)
+            sql_field = determine(self.compute_sql, model, table)
             assert isinstance(sql_field, SQL), f"{self} invalid return of compute_sql"
             return sql_field
         if not self.store or not self.column_type:
             if self.related and not self.store:
-                model, field, alias = self.traverse_related_sql(model, alias, query)
-                return model._field_to_sql(alias, field.name, query)
+                # traverse_related
+                if self.compute_sudo:
+                    table = table._sudo()
+                traverse = table
+                for fname in self.related.split('.'):
+                    traverse = traverse[fname]
```

Comment:
When traversing an x2many field, this calls `field.property_to_sql(...)`, which should raise some exception.  Simply make sure that the exception is comprehensible for a developer ("not implemented" is not very helpful.)